Update h5py

thomashampson opened the pull request:
**Note to gatekeepers: Please also merge mantidproject/conda-recipes#89 when this is merged.**

**Description of work.**

Remove ancient h5py/hdf5 pin and move to HDF5 1.12/h5py.

**To test:**

* Code review and all automated tests should pass. HDF5 is core to all NeXus loading so any failures would be picked up by the wealth of tests that load/save files.

Fixes #34731 

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

thomashampson committed:
Remove h5py and hdf5 pinning

sf1919 committed:
Fix issues with unpinning h5py

Co-authored-by: jhaigh0 <jonathan.haigh@stfc.ac.uk>
Co-authored-by: thomashampson <thomas.hampson@stfc.ac.uk>

martyngigg commented:
I ran this locally on my Windows machine. I got a subset of the failures:
```
The following tests FAILED:
        1829 - MDAlgorithmsTest_CloneMDWorkspaceTest (Failed)
        1899 - MDAlgorithmsTest_MinusMDTest (Failed)
        1903 - MDAlgorithmsTest_PlusMDTest (Failed)
        1912 - MDAlgorithmsTest_SaveMDTest (Failed)
        1920 - MDAlgorithmsTest_TransformMDTest (Failed)
```
All of them some file access error. I don't see this issue with the clashing versions though. I wonder if mamba hasn't uninstalled the old hdf5 properly?

martyngigg committed:
Replace new/delete with unique_ptr in test

martyngigg committed:
Fix copying file backend if HDF5 has locked the file

Version 1.12 of HDF5 in Conda configured file locking
on access for the first time. If a file is open on Windows
it has an exclusive lock and a read handle cannot be
obtained to copy the file. The workaround is to close
the file, copy then reopen in the same mode.

martyngigg committed:
Pin HDF5 to version 1.12

New HDF5 releases that are not patches general require some
changes to compile.
Pin to current major.minor to avoid new versions being picked
up immediately. The pin will allow patch updates.

martyngigg committed:
Add a release note for HDF5 upgrade
